### ISSUE TITLE
Some Iosevka variants broken

### ISSUE THREAD

**Issue by Finii** [ryanoasis/nerd-fonts#694]
#### 🗹 Requirements

- [x] A brief but descriptive _title_ of your issue
- [x] I have searched the [issues](https://github.com/ryanoasis/nerd-fonts/issues) for my issue and found nothing related and/or helpful
- [x] I have read or scanned the [FAQ](https://github.com/ryanoasis/nerd-fonts/wiki/FAQ-and-Troubleshooting)
- [x] I have read or scanned the [Wiki](https://github.com/ryanoasis/nerd-fonts/wiki)

## The patched Iosevka Mono fonts are more or less broken

Some Iosevka Mono fonts can not be opened at all (or it takes more than some minutes), while most other open with error messages of this kind:

```
patched-fonts/Iosevka/Thin-Italic/complete/Iosevka Thin Italic Nerd Font Complete Mono.ttf
PythonUI_Init()
copyUIMethodsToBaseTable()
Internal Error:
Subtable status not filled in for 0th subtable of 'calt' Contextual Alternates lookup 26
Subtable status not filled in for 1th subtable of 'calt' Contextual Alternates lookup 26
Subtable status not filled in for 2th subtable of 'calt' Contextual Alternates lookup 26
Subtable status not filled in for 0th subtable of 'dlig' Discretionary Ligatures lookup 32
Subtable status not filled in for 1th subtable of 'dlig' Discretionary Ligatures lookup 32
Subtable status not filled in for 2th subtable of 'dlig' Discretionary Ligatures lookup 32
Subtable status not filled in for 0th subtable of Reverse Contextual Chaining Substitution lookup 35
Subtable status not filled in for 1th subtable of Reverse Contextual Chaining Substitution lookup 35
Subtable status not filled in for 2th subtable of Reverse Contextual Chaining Substitution lookup 35
Subtable status not filled in for 0th subtable of Reverse Contextual Chaining Substitution lookup 38
Subtable status not filled in for 1th subtable of Reverse Contex
```

I believe fontforge tries to fix the fonts that do not open for extended times like this:
```
...
Attempt to reference lookup 11293 (within a contextual lookup), but there are only 429 lookups in 'GSUB'
Attempt to reference lookup 11805 (within a contextual lookup), but there are only 429 lookups in 'GSUB'
Attempt to reference lookup 12317 (within a contextual lookup), but there are only 429 lookups in 'GSUB'
Attempt to reference lookup 14109 (within a contextual lookup), but there are only 429 lookups in 'GSUB'
Attempt to reference lookup 14877 (within a contextual lookup), but there are only 429 lookups in 'GSUB'
Attempt to reference lookup 15901 (within a contextual lookup), but there are only 429 lookups in 'GSUB'
Attempt to reference lookup 23837 (within a contextual lookup), but there are only 429 lookups in 'GSUB'
Attempt to reference lookup 24605 (within a contextual lookup), but there are only 429 lookups in 'GSUB'
Attempt to reference lookup 25373 (within a contextual lookup), but there are only 429 lookups in 'GSUB'
Attempt to reference lookup 25885 (within a contextual lookup), but there are only 429 lookups in 'GSUB'
Attempt to reference lookup 27933 (within a contextual lookup), but there are only 429 lookups in 'GSUB'
Attempt to reference lookup 28701 (within a contextual lookup), but there are only 429 lookups in 'GSUB'
Attempt to reference lookup 47901 (within a contextual lookup), but there are only 429 lookups in 'GSUB'
Attempt to reference lookup 49181 (within a contextual lookup), but there are only 429 lookups in 'GSUB'
...
[on and on is scrolls]
```

**_I guess the problem is the ligature removal_** with the `src/unpatched-fonts/Iosevka/config.json`.

#### 🔧 Your Setup

- _Which font are you using (e.g. `Anonymice Powerline Nerd Font Complete.ttf`)?_ Iosevka Mono *
- _Which terminal emulator are you using (e.g. `iterm2`, `urxvt`, `gnome`, `konsole`)?_
none, `fontforge` (20201107) to just open the font
Used both:
```
 Based on sources from 2021-01-15 15:55 UTC-ML-D-GDK3.
 Based on sources from 2021-11-29 12:07 UTC-ML-D-GDK3. (HEAD)
```
- _Are you using OS X, Linux or Windows? And which specific version or distribution?_ Ubuntu 21.10

#### ★ Optional

Maybe these are related:
#585 #586 

![image](https://user-images.githubusercontent.com/16012374/143863849-1818fca7-e23f-462d-93d9-2935ab16eec0.png)

**Comment by Finii** [comment]
An especially severe problem have these two fonts:

```
iosevka-heavyoblique
iosevka-term-heavyoblique
```

After (any, even manual) patching, the patched result can not even be opened with `fontforge`(*).

I even tried the newest [release v11.2.0](https://github.com/be5invis/Iosevka/releases/tag/v11.2.0).

This has nothing to do with the ligature removal (opposed to what I wrote above).

Note the error message in the bottom:

![image](https://user-images.githubusercontent.com/16012374/145577914-401ea700-6878-4f43-a04a-34fa41476f73.png)


(*) Opened programatically via Python. It opens in interactive `fontforge` after a long long wait.

**Comment by GulajavaMinistudio** [comment]
I'm also facing same issue. I patch [Iosevka 11.2.3](https://github.com/be5invis/Iosevka/releases/tag/v11.2.3) and want to keep built in ligature on Iosevka . Successfully patched with some error warning like this. **"Internal error: Internal Error: Attempt to output 66740 into a 16-bit field. It will be truncated and the file may not be useful."**

![Screen Shot 2021-12-29 at 07 34 53](https://user-images.githubusercontent.com/11639412/147620474-be3f97f6-d2d5-43f2-a71d-013bec5e9665.png)

And after install it on Linux Ubuntu, and try in VS Code text editor, the patched font didn't work properly. It not show ligature. But if i use unpatched Iosevka font, it work properly.

**Comment by Finii** [comment]
Hmm, at this point `fontforge` creates the message (inside `putshort()`) (i.e. this is the caller).

![image](https://user-images.githubusercontent.com/16012374/147787274-b1488b86-cd45-4833-a4a8-2122ff8e4705.png)

**Comment by Finii** [comment]
Well, it has nothing to do with our patching.

While I can open `src/unpatched-fonts/Iosevka/Regular/iosevka-term-regular.ttf ` and generate a new unchanged font **in `fontforge` interatively**, I can not do that in Python mode:

![image](https://user-images.githubusercontent.com/16012374/147910911-f71c995e-6e77-4f47-9643-08965f2b9253.png)

**Comment by Finii** [comment]
WHAT THE F***!

It does save correctly if we supply an empty flag tuple (instead of no flags):

```python
#!/usr/bin/env python
import fontforge

font = fontforge.open('src/unpatched-fonts/Iosevka/Regular/iosevka-term-regular.ttf')
font.generate('iosevka_blah.ttf', flags=tuple())
```

:sob:

**Comment by Finii** [comment]
```c
/* Legacy screw ups mean that opentype & apple bits don't mean what */
/*  I want them to. Python users should not see that, but fix it up */
/*  here */
if ( (iflags&0x80) && (iflags&0x10) )   /* Both */    
    iflags &= ~0x10;
else if ( (iflags&0x80) && !(iflags&0x10)) /* Just opentype */
    iflags &= ~0x80;
else if ( !(iflags&0x80) && (iflags&0x10)) /* Just apple */
    /* This one's set already */;
else
    iflags |= 0x90;
    
[...]

                fm_flag_apple = 0x10,
                fm_flag_opentype = 0x80,
```

If we set "apple" and "opentype" -> `fm_flag_opentype`
If we set "opentype" -> `0`
If we set "apple" -> `fm_flag_apple`
If we set nothing -> `fm_flag_apple` | `fm_flag_opentype`

**Comment by Finii** [comment]
If we set no flags by setting an empty tuple the correction will happen and we in fact set `apple | opentype`.

That will later be translated here:
```c
old_sfnt_flags = 0;
switch ( fmflags&(fm_flag_apple|fm_flag_opentype) ) {
  case fm_flag_opentype:
    old_sfnt_flags |= ttf_flag_applemode|ttf_flag_otmode;
  break;
  case fm_flag_apple|fm_flag_opentype:
    /* Neither */;
  break;
  case fm_flag_apple:
    old_sfnt_flags |= ttf_flag_applemode;
  break;
  case 0x00:
    old_sfnt_flags |= ttf_flag_otmode;
  break;
  default:
  break;
}
```
If we set "apple" and "opentype" -> `fm_flag_opentype` -> `ttf_flag_applemode | ttf_flag_otmode`
If we set "opentype" -> `0` -> `ttf_flag_otmode`
If we set "apple" -> `fm_flag_apple` -> `ttf_flag_applemode`
If we set nothing -> `fm_flag_apple | fm_flag_opentype` -> `0`


If we set no flags by ... just not mentioning, we use `-1` as flags and the correction does never happen (instead we use "Default to what we did last time" :grimacing:)
Which means
```c
int old_sfnt_flags = ttf_flag_otmode;
```

**Comment by Finii** [comment]
Interactive mode (default) works, because it uses `apple` mode and not `opentype`.

![image](https://user-images.githubusercontent.com/16012374/147948344-26d26925-9a90-4ed7-aefd-059eaa26c222.png)

It fails with the same error if `opentype` is selected.
So I would say this is an `fontforge` error, and we can not really do something about it (apart from fixing upstream).

**Comment by Finii** [comment]
Fixed via #884
